### PR TITLE
ARROW-2184: [C++]  Add static constructor for FileOutputStream returning shared_ptr to OutputStream

### PR DESCRIPTION
**Pull request by xuepanchen**
Add constructors to return pointers to the base interface

### CODE REVIEW

**Comment by wesm** [comment]
Just rebased this, please 

```
git fetch origin
git reset --hard origin/ARROW-2184
```

**Review by xhochy — APPROVED** [review]
+1, LGTM.

I would like to keep the more specific ones as they are nice to have if you work in a more strongly typed, pure C++ environment.

File: cpp/src/arrow/io/file.h [hunk 1]
```diff
@@ -39,6 +39,21 @@ class ARROW_EXPORT FileOutputStream : public OutputStream {
...
+  /// When opening a new file, any existing file with the indicated path is
+  /// truncated to 0 bytes, deleting any existing memory
+  static Status Open(const std::string& path, std::shared_ptr<OutputStream>* out);
+
+  /// \brief Open a local file for writing
+  /// \param[in] path with UTF8 encoding
+  /// \param[in] append append to existing file, otherwise truncate to 0 bytes
+  /// \param[out] out a base interface OutputStream instance
+  static Status Open(const std::string& path, bool append,
+                     std::shared_ptr<OutputStream>* out);
```
Comment (wesm) [inline cpp/src/arrow/io/file.h]: We need some unit tests for these. We should also change the Python bindings to use these APIs

Comment (xuepanchen) [reply cpp/src/arrow/io/file.h]: Do we need to change the Python bindings in this pull request?

Comment (wesm) [reply cpp/src/arrow/io/file.h]: I pushed a commit here to make the change, so we're good now

**Comment by wesm** [comment]
+1. Will merge once the build is green. Thanks!

**Review by xhochy — APPROVED** [review]

**Pull request closed without merging**